drivers/flash/soc_flash_numaker.c

```diff
@@ -0,0 +1,281 @@
+				rc = -EIO;
+				goto done;
+			}
+			page_nums -= (FMC_BANK_SIZE / FMC_FLASH_PAGE_SIZE);
+			addr += FMC_BANK_SIZE;
+		} else {
+			/* erase page */
+			if (FMC_Erase(addr)) {
+				LOG_ERR("Erase flash page failed or erase time-out");
+				rc = -EIO;
+				goto done;
+			}
+			page_nums--;
+			addr += FMC_FLASH_PAGE_SIZE;
+		}
+	}
+
+done:
+	SYS_LockReg();
+	irq_unlock(key);
+	/* release semaphore */
+	k_sem_give(&dev_data->write_lock);
+
+	return rc;
+}
+
+/*
+ * Read a flash memory area.
+ *
+ * param dev       Device struct
+ * param offset    The address's offset
+ * param data      The buffer to store or read the value
+ * param length    The size of the buffer
+ * return 0       on success,
+ * return -EIO     erroneous code
+ */
+static int flash_numaker_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* Validate range */
+	if (!flash_numaker_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* read flash */
+	memcpy(data, (void *)addr, len);
+
+	return 0;
+}
+
+static int32_t flash_numaker_block_write(uint32_t u32_addr, uint8_t *pu8_data, int block_size)
+{
+	int32_t retval;
+	uint32_t *pu32_data = (uint32_t *)pu8_data;
+
+	SYS_UnlockReg();
+	if (block_size == 4) {
+		retval = FMC_Write(u32_addr, *pu32_data);
+	} else if (block_size == 8) {
+		retval = FMC_Write8Bytes(u32_addr, *pu32_data, *(pu32_data + 1));
+	} else {
+		retval = -1;
+	}
+	SYS_LockReg();
+
+	return retval;
+}
+
+static int flash_numaker_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t rc = 0;
+	unsigned int key;
+	uint32_t addr = dev_data->flash_block_base + offset;
+	int block_size = flash_numaker_parameters.write_block_size;
+	int blocks = (len / flash_numaker_parameters.write_block_size);
+	uint8_t *pu8_data = (uint8_t *)data;
+	uint8_t buffer[SOC_NV_FLASH_WRITE_BLOCK_SIZE];
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* Validate range */
+	if (!flash_numaker_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* Validate address alignment */
+	if ((addr % flash_numaker_parameters.write_block_size) != 0) {
+		return -EINVAL;
+	}
+
+	if (k_sem_take(&dev_data->write_lock, K_FOREVER)) {
+		return -EACCES;
+	}
+
+	key = irq_lock();
+
+	while (blocks) {
+		if (flash_numaker_block_write(addr, pu8_data, block_size)) {
+			rc = -EIO;
+			goto done;
+		}
+		pu8_data += block_size;
+		addr += block_size;
+		blocks--;
+	}
+	if ((len % block_size) != 0) {
+		memcpy(buffer, (void *)addr, block_size);
+		memcpy(buffer, (void *)pu8_data, len / block_size);
+		if (flash_numaker_block_write(addr, buffer, block_size)) {
+			rc = -EIO;
+			goto done;
+		}
+	}
```

Comment:
Writes unaligned to `write_block_size` are not allowed, write callback should not correct user:
https://docs.zephyrproject.org/apidoc/latest/group__flash__interface.html#ga76d7880cc5e18ca40237736d3bd94324

Reply:
Revise the flash_numaker_write() based on your advice & flash_write spec.
Update at commit  e8777cd8a83f